src/batch.rs

```diff
@@ -91,64 +138,96 @@ pub enum BatchStatus {
+                    ..Default::default()
+                };
+                Self::parse_response_file(file, client).await?
+            }
+        };
+        Ok(results)
+    }
+
+    async fn from_operation(
+        operation: BatchOperation,
+        client: Arc<GeminiClient>,
+    ) -> Result<Self, Error> {
         if operation.done {
             // According to Google API documentation, when done=true, result must be present
             let result = operation.result.context(MissingResultSnafu {
                 name: operation.name.clone(),
             })?;
 
-            match result {
-                OperationResult::Failure { error } => Err(error).context(BatchFailedSnafu {
-                    name: operation.name,
-                }),
-                OperationResult::Success { response } => {
-                    let mut results: Vec<BatchResultItem> = response
-                        .inlined_responses
-                        .inlined_responses
-                        .into_iter()
-                        .map(|item| match item {
-                            BatchGenerateContentResponseItem::Success { response, metadata } => {
-                                BatchResultItem::Success {
-                                    key: metadata.key,
-                                    response,
-                                }
-                            }
-                            BatchGenerateContentResponseItem::Error { error, metadata } => {
-                                BatchResultItem::Error {
-                                    key: metadata.key,
-                                    error,
-                                }
-                            }
-                        })
-                        .collect();
-
-                    // Sort results by key to ensure a consistent order.
-                    results.sort_by_key(|item| {
-                        let key_str = match item {
-                            BatchResultItem::Success { key, .. } => key,
-                            BatchResultItem::Error { key, .. } => key,
-                        };
-                        key_str.parse::<usize>().unwrap_or(usize::MAX)
-                    });
-
-                    // Handle terminal states based on metadata for edge cases
-                    match operation.metadata.state {
-                        BatchState::BatchStateCancelled => Ok(BatchStatus::Cancelled),
-                        BatchState::BatchStateExpired => Ok(BatchStatus::Expired),
-                        _ => Ok(BatchStatus::Succeeded { results }),
-                    }
-                }
+            let response = Result::from(result).context(BatchFailedSnafu {
+                name: operation.name,
+            })?;
+
+            let mut results = Self::process_successful_response(response, client).await?;
+            results.sort_by_key(|r| r.meta.key.parse::<usize>().unwrap_or(usize::MAX));
```

Comment:
https://ai.google.dev/api/batch-mode#GenerateContentBatchOutput

> The responses will be in the same order as the input requests.

This line from the Google documentation turned out to be misleading: if the returned values are not sorted, their order will not match the order from the original request.

I will create issue for Google. Based on my past experience, I don't expect them to fix this bug (or documentation quickly).

This approach may not work correctly for batches created outside of this crate (as there's no guarantee that the numeric identifiers can be parsed), but I'm not sure where we should mention this.